File: DP(Dynamic_Programming, 동적계획법)/백준_11727_2xn타일링2/2xn타일링2.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int arr[1000] = { 0,1,3 };

int main() {
	int n = 0;
	scanf("%d", &n);

	for (int i = 3; i <= n; i++) {
		int m = arr[i - 1] * 2;
		(i % 2 == 0) ? m += 1 : m -= 1;
		arr[i] = (m >= 10007) ? m % 10007 : m;
	}
	printf("%d", arr[n]);


}

File: DP(Dynamic_Programming, 동적계획법)/백준_11727_2xn타일링2/2xn타일링2.py
arr = [0 for _ in range(1001)]
arr[0] = 0;
arr[1] = 1;
arr[2] = 3;
n = int(input())
for i in range(3, n+1):
    arr[i] = arr[i-1] * 2
    if i % 2 == 0: arr[i] += 1
    else: arr[i]-=1
    arr[i] %=10007
print(arr[n])